rllib/policy/torch_policy_v2.py

```diff
@@ -0,0 +1,1220 @@
+    def get_num_samples_loaded_into_buffer(self, buffer_index: int = 0) -> int:
+        if len(self.devices) == 1 and self.devices[0] == "/cpu:0":
+            assert buffer_index == 0
+        return len(self._loaded_batches[buffer_index])
+
+    @override(Policy)
+    @DeveloperAPI
+    def learn_on_loaded_batch(self, offset: int = 0, buffer_index: int = 0):
+        if not self._loaded_batches[buffer_index]:
+            raise ValueError(
+                "Must call Policy.load_batch_into_buffer() before "
+                "Policy.learn_on_loaded_batch()!"
+            )
+
+        # Get the correct slice of the already loaded batch to use,
+        # based on offset and batch size.
+        device_batch_size = self.config.get(
+            "sgd_minibatch_size", self.config["train_batch_size"]
+        ) // len(self.devices)
+
+        # Set Model to train mode.
+        if self.model_gpu_towers:
+            for t in self.model_gpu_towers:
+                t.train()
+
+        # Shortcut for 1 CPU only: Batch should already be stored in
+        # `self._loaded_batches`.
+        if len(self.devices) == 1 and self.devices[0].type == "cpu":
+            assert buffer_index == 0
+            if device_batch_size >= len(self._loaded_batches[0][0]):
+                batch = self._loaded_batches[0][0]
+            else:
+                batch = self._loaded_batches[0][0][offset : offset + device_batch_size]
+            return self.learn_on_batch(batch)
+
+        if len(self.devices) > 1:
+            # Copy weights of main model (tower-0) to all other towers.
+            state_dict = self.model.state_dict()
+            # Just making sure tower-0 is really the same as self.model.
+            assert self.model_gpu_towers[0] is self.model
+            for tower in self.model_gpu_towers[1:]:
+                tower.load_state_dict(state_dict)
+
+        if device_batch_size >= sum(len(s) for s in self._loaded_batches[buffer_index]):
+            device_batches = self._loaded_batches[buffer_index]
+        else:
+            device_batches = [
+                b[offset : offset + device_batch_size]
+                for b in self._loaded_batches[buffer_index]
+            ]
+
+        # Callback handling.
+        batch_fetches = {}
+        for i, batch in enumerate(device_batches):
+            custom_metrics = {}
+            self.callbacks.on_learn_on_batch(
+                policy=self, train_batch=batch, result=custom_metrics
+            )
+            batch_fetches[f"tower_{i}"] = {"custom_metrics": custom_metrics}
+
+        # Do the (maybe parallelized) gradient calculation step.
+        tower_outputs = self._multi_gpu_parallel_grad_calc(device_batches)
+
+        # Mean-reduce gradients over GPU-towers (do this on CPU: self.device).
+        all_grads = []
+        for i in range(len(tower_outputs[0][0])):
+            if tower_outputs[0][0][i] is not None:
+                all_grads.append(
+                    torch.mean(
```

Comment:
hmm, wait, torch.mean([g1, g2, g3]) is the average gradient right?
the scale of the gradient should be the same? it's like manual all-reduce.
maybe I am missing something.

Reply:
Let's say we have 3 state/action samples that produce losses `0.1, 0.2, 0.3`. If we were to train all three samples using a single tower, the loss would be `loss = 0.1 + 0.2 + 0.3 = 0.6`. However, now let's say we compute the same exact losses using three towers. If we take the mean rather than sum: `loss = (0.1 + 0.2 + 0.3) / 0.3 = 0.2`.

So if we take the mean, then increasing the number of towers will actually "slow" down the learning. So if you change the number of towers, you would need to multiply the `lr` by `num_towers` to retain the same timestep efficiency.

Reply:
> Let's say we have a rollout of three timesteps that produces losses 0.1, 0.2, 0.3. If we were to train the entire rollout using a single tower, the loss would be loss = 0.1 + 0.2 + 0.3 = 0.6.

The mostly used reduction method for training losses is `mean`. Each tower loss is averaged over the number of samples (`batch_size * seq_lens`). If we compute the loss using a single tower, the loss is also `loss = (0.1 + 0.2 + 0.3) / 3 = 0.2`.

On each tower, we reduce losses over the number of samples:

https://github.com/ray-project/ray/blob/de69b0d6d625615a58ac5a80d25f0cf81882b0f4/rllib/agents/ppo/ppo_torch_policy.py#L107-L110

https://github.com/ray-project/ray/blob/de69b0d6d625615a58ac5a80d25f0cf81882b0f4/rllib/agents/ppo/ppo_torch_policy.py#L136-L143

Reply:
In that case, then by taking the mean over towers we are taking a mean of means, which is different than the mean over the number of samples.

Reply:
> In that case, then by taking the mean over towers we are taking a mean of means, which can be different than the mean over the number of samples if the number of samples varies between towers.

If the batch is evenly divided for all towers, the result of "mean" vs. "mean of mean" are identical. The condition is always true for non-RNN cases, but is not always fulfilled when we have `seq_lens`  for RNNs.

Reply:
Hey @smorad , @XuehaiPan , and @gjoliver and everyone else reading this.
I completely agree with Steven's "scaredness" about our Policy classes. The amount and complexity of the different methods that one could/should/must-not override is insane. I'm prepping a WIP PR here that addresses this problem:
https://github.com/ray-project/ray/pull/25293

This will have to be iterated over and is not the final solution.